### PR TITLE
Conduit: Exporter daemon integration

### PR DESCRIPTION
**Pull request by Eric-Warehime**
## Summary

The goal of this PR is to integrate the new Exporter code with the existing Indexer daemon. In this way we can enable users to create and plug new data destinations into the existing Indexer framework.
 
Replaces #1134 since there have been a lot of changes which are easier to just branch off of instead of merge into the old PR.

## Test Plan

Added tests to cover new code.

### CODE REVIEW

File: exporters/noop/noop_exporter_test.go [hunk 1]
```diff
@@ -16,18 +16,18 @@ var ne = nc.New()
...
 
 func TestExporterMetadata(t *testing.T) {
 	meta := ne.Metadata()
 	assert.Equal(t, plugins.PluginType(plugins.Exporter), meta.Type())
-	assert.Equal(t, noopExporterMetadata.ExpName, meta.Name())
-	assert.Equal(t, noopExporterMetadata.ExpDescription, meta.Description())
-	assert.Equal(t, noopExporterMetadata.ExpDeprecated, meta.Deprecated())
+	assert.Equal(t, Metadata.ExpName, meta.Name())
+	assert.Equal(t, Metadata.ExpDescription, meta.Description())
+	assert.Equal(t, Metadata.ExpDeprecated, meta.Deprecated())
```
Comment (winder) [inline exporters/noop/noop_exporter_test.go:30]: Seems like we could use the same Metadata object in Export/Importer until we need them to diverge.

File: processor/processor.go [hunk 2]
```diff
@@ -1,13 +1,12 @@
...
-	"github.com/algorand/go-algorand/ledger/ledgercore"
 	"github.com/algorand/go-algorand/rpcs"
 )
 
 // Processor is the block processor interface
 type Processor interface {
 	Process(cert *rpcs.EncodedBlockCert) error
-	SetHandler(handler func(block *ledgercore.ValidatedBlock) error)
 	NextRoundToProcess() uint64
+	AddGenesisBlock() error
```
Comment (winder) [inline processor/processor.go:11]: Is this going away in Stephen's PR? Seems like a weird method.

**Review by winder — COMMENTED** [review]
I'm not sure how useful this feedback is since we have a lot of conflicts with #1140.

File: cmd/algorand-indexer/daemon.go [hunk 3]
```diff
@@ -315,61 +324,86 @@ func runDaemon(daemonConfig *daemonConfig) error {
...
 }
 
-func runBlockImporter(ctx context.Context, cfg *daemonConfig, wg *sync.WaitGroup, db idb.IndexerDb, dbAvailable chan struct{}, bot fetcher.Fetcher, opts idb.IndexerDbOptions) {
+// Bridges functionality between setting config via CLI and the new plugin configs
+func writeCLIToPlugin(cfg *daemonConfig) {
+	switch cfg.exporterPlugin {
+	case postgresql.Metadata.Name():
+		pluginCfg := plugins.LoadConfig(logger, cfg.indexerDataDir, postgresql.Metadata)
+		var expCfg postgresql.ExporterConfig
+		err := yaml.Unmarshal([]byte(pluginCfg), &expCfg)
```
Comment (shiqizng) [inline cmd/algorand-indexer/daemon.go:373]: nit: maybe use this pattern
```suggestion
		if err := yaml.Unmarshal([]byte(pluginCfg), &expCfg); err != nil {
			maybeFail(err, "unable to load postgresql config: %v", err)
		}
```

File: exporters/postgresql/postgresql_exporter.go [hunk 4]
```diff
@@ -13,7 +13,8 @@ import (
 	"gopkg.in/yaml.v3"
 )
 
-const exporterName = "postgresql"
+// ExporterName constant for the postgresql Exporter module
+const ExporterName = "postgresql"
```
Comment (shiqizng) [inline exporters/postgresql/postgresql_exporter.go:17]: why does this need to be exported?

File: processor/blockprocessor/block_processor.go [hunk 5]
```diff
@@ -3,146 +3,64 @@ package blockprocessor
...
 	if l == nil {
 		return nil, fmt.Errorf("MakeProcessorWithLedger() err: local ledger not initialized")
 	}
-	err := addGenesisBlock(l, handler)
+	var proc processor.Processor
+	if exp, ok := genericHandler.(exporters.Exporter); ok {
+		proc = &ledgerExporterProcessor{logger: logger, ledger: l, exporter: exp}
+	} else if handler, ok := genericHandler.(func(block *ledgercore.ValidatedBlock) error); ok {
+		proc = &blockProcessor{logger: logger, ledger: l, handler: handler}
+	} else if genericHandler == nil {
```
Comment (shiqizng) [inline processor/blockprocessor/block_processor.go:32]: I think if genericHandler == nil the if statements before this one will fail with an error? I don't believe we are allowed to case nil to a type.

File: processor/blockprocessor/exporter_processor.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+		})
+		if err != nil {
+			return fmt.Errorf("addGenesisBlock() handler err: %w", err)
+		}
+	}
+	return nil
+}
+
+// Process a raw algod block
+func (proc *ledgerExporterProcessor) Process(blockCert *rpcs.EncodedBlockCert) error {
```
Comment (shiqizng) [inline processor/blockprocessor/exporter_processor.go:43]: it looks like the same logic is already ran in [block_processor,](https://github.com/algorand/indexer/blob/45d5c4b28523098696f438d98602ad411f84832e/processors/blockprocessor/block_processor.go#L127) why do we need this in the exporter again?

File: processor/blockprocessor/handler_processor.go [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+		err = proc.handler(&vb)
+		if err != nil {
+			return fmt.Errorf("addGenesisBlock() handler err: %w", err)
+		}
+	}
+	return nil
+}
+
+// Process a raw algod block
+func (proc *blockProcessor) Process(blockCert *rpcs.EncodedBlockCert) error {
```
Comment (shiqizng) [inline processor/blockprocessor/handler_processor.go:36]: I may be missing something. why do I seem to see similar Process logic repeated in multiple places?

File: cmd/algorand-indexer/daemon.go [hunk 8]
```diff
@@ -315,61 +324,86 @@ func runDaemon(daemonConfig *daemonConfig) error {
...
 	}
 
-	fmt.Printf("serving on %s\n", daemonConfig.daemonServerAddr)
-	logger.Infof("serving on %s", daemonConfig.daemonServerAddr)
-
-	options := makeOptions(daemonConfig)
-
-	api.Serve(ctx, daemonConfig.daemonServerAddr, db, bot, logger, options)
+	// Only serve the API for the postgres exporter
+	if daemonConfig.exporterPlugin == postgresql.ExporterName {
```
Comment (shiqizng) [inline cmd/algorand-indexer/daemon.go:353]: is Postgres or postGres considered different from postgres?

File: cmd/algorand-indexer/daemon.go [hunk 9]
```diff
@@ -315,61 +324,86 @@ func runDaemon(daemonConfig *daemonConfig) error {
...
+		expCfg.ConnectionString = postgresAddr
+		data, err := yaml.Marshal(expCfg)
+		if err != nil {
+			maybeFail(err, "failure setting postgresql config data: %v", err)
+		}
+		err = ioutil.WriteFile(plugins.PluginConfigPath(cfg.indexerDataDir, postgresql.Metadata), []byte(data), fs.ModePerm)
+		if err != nil {
+			maybeFail(err, "failure writing updated options to postgresql exporter config file: %v", err)
+		}
+	default:
```
Comment (shiqizng) [inline cmd/algorand-indexer/daemon.go:388]: maybe in default case write to logger that no matching cfg.exporterPlugin was found?

**Comment by Eric-Warehime** [comment]
Cancelling this PR--I've taken a look at all of the feedback and will use it to inform how I tackle the integration between conduit and indexer's daemon. But based on the fact that we have created a new conduit binary on our feature branch, I don't think it makes sense to implement plugin types as individual features in daemon instead of modeling the indexer as a conduit pipeline with an API wrapper.

**Pull request closed without merging**